### ISSUE TITLE
Is my code correct for tensor fusion?

### ISSUE THREAD

**Issue by convman** [A2Zadeh/CMU-MultimodalSDK#136]
def tensor_fuse(out):
    o1 = out[0]     # o1.shape = (None,32)
    o2 = out[1]     # o2.shape = (None,32)
    o3 = out[2]     # o3.shape = (None,32)
  
    x = tf.expand_dims(o1,1)   # x.shape = (None,1,32)
    y = tf.expand_dims(o2,2)   # y.shape = (None,32,1)
    o12 = tf.reshape(tf.multiply(x,y),shape=[-1,32*32])  # o12.shape = (None,32*32)
  
    x = tf.expand_dims(o2,1)   # x.shape = (None,1,32)
    y = tf.expand_dims(o3,2)   # y.shape = (None,32,1)
    o23 = tf.reshape(tf.multiply(x,y),shape=[-1,32*32])  # o23.shape = (None,32*32)
  
    x = tf.expand_dims(o3,1)   # x.shape = (None,1,32)
    y = tf.expand_dims(o1,2)   # y.shape = (None,32,1)
    o31 = tf.reshape(tf.multiply(x,y),shape=[-1,32*32])  # o31.shape = (None,32*32)
  
    x = tf.expand_dims(o12,2)  # x.shape = (None,32*32,1)
    y = tf.expand_dims(o3,1)   # y.shape = (None,1,32)
    o123 = tf.reshape(tf.multiply(x,y),[-1,32*32*32])  # o123.shape = (None,32*32*32)

    return concatenate([o1,o2,o3,o12,o23,o31,o123])

What I could understand with tensor fusion: was to find outer product between different modalities and concatenate them. After then, feeding them to DNN, use prediction layer. Please correct me if I am wrong. Provided code is to apply tensor-fusion in keras,Please tell me if its correct?

**Comment by A2Zadeh** [comment]
Hi @convman, quick questions, is this for Keras? If not have you considered using TFN in mmmodelsdk?

**Comment by convman** [comment]
Yes its for keras...want to integrate it with rest of my code. I have not used mmmodelsdk yet.

**Comment by A2Zadeh** [comment]
The code is correct, at least from what I can tell by looking at it (I don't know the dims of inputs). Alternatively, you could just extend each modality by a single constant and achieve the same thing. So basically say you have the following dims:

language: bs x 128
vision: bs x 64
acoustic bs x 32

then simply add a 1 to each one of their dimensions to get:

language: bs x 129
vision: bs x 65
acoustic bs x 33

And then do the outer product.

**Comment by convman** [comment]
yes i get that. My input feature shape for all modalities is (None, 32)- None for batch size.
Thank you!

**Comment by A2Zadeh** [comment]
Great. Then just pad each modality by a constant (no grad needed for it). And then do outer-product. No need to build each subtensor individually.

**Comment by convman** [comment]
Actually, I faced difficulty in padding and doing outer product in keras layers that's why I resolved to this solution. I hope it still does the same job as the original code. 
Thank you for the feedback!

**Comment by A2Zadeh** [comment]
@convman sounds good! Let me know how it goes! BTW, pull the SDK again and get updated gloves for CMU-MOSEI and CMU-MOSI. I just updated them.
